Fix data_source and data_source_async bugs

wm75 opened the pull request:
Currently, the behavior of tools of type `data_source` and `data_source_async` deviates from each other and from documented behavior. Here, I'll try to fix at least some of the following issues:

- [x] though undocumented, `data_source` tools that declare profile 21.09 or higher will have their python env isolated from Galaxy's. However, this was never implemented for `data_source_async` tools
- [x] `data_source_async` tools ignore their https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-request-param-translation section
- [x] `data_source_async` tools switch to error state when receiving a non-OK response from the remote server to their second request, but continue executing and fetching data, and when succesful flip their state to OK.
- [x] the default `data_source.py` script in `tools/data_source/` is full of historic cruft and currently only supports legacy tool-provided metadata.
- [x] shouldn't leak internal data through POST requests

While I'm working on this, feel free to comment on individual commits :-)

## How to test the changes?
You can test changes manually by playing with https://github.com/wm75/galaxy-data_source-examples/tree/main/cherrypy.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

wm75 committed:
Isolate python env of data_source_async tools

and document profile-dependency.

mvdbeek commented on lib/galaxy/tools/__init__.py:
```diff
@@ -910,13 +910,13 @@ def requires_galaxy_python_environment(self):
             return True
 
-        if self.tool_type == "data_source" and self.profile < 21.09:
+        if self.tool_type in ["data_source", "data_source_async"] and self.profile < 21.09:
```
Can you split that out and make it self.profile < 24.0 ?

wm75 replied:
Sure, if you think that's the safer option.

wm75 committed:
Shift data_source_async env isolation to 24.0 profile

wm75 committed:
Preserve request params declared by data_source_async tools

wm75 committed:
Fix linting

wm75 committed:
Refactor async params mapping

wm75 committed:
Prevent updating of terminal datasets

Async data requests coming in after the corresponding dataset has been
finalized are now ignored.

wm75 committed:
Fix linting

wm75 committed:
Fix linting once more

wm75 committed:
Start modernizing data_source.py script

- Drop support for very old format of specifying execution params in a
  tabular file instead of a json one
- Use O_TRUNC to discard previous file contents instead of opening for
  writing, then closing
- Do not silently ignore non-integer values for max_file_size

wm75 committed:
Simply exit on errors

wm75 committed:
Switch data_source.py to modern tool-provided metadata

Allows moving all data source tools using the script to profile 20.09.
For more recent profiles data_source.py would have to work outside of
Galaxy's Python environment.

wm75 committed:
Fix linting

wm75 committed:
Try to fix API test

The test tool needs to declare a recent enough profile to expect modern
tool-provided metadata.

wm75 committed:
Format import section with isort

wm75 committed:
Preserve original request params as nested dict

and use them in final POST request

wm75 committed:
Fix linting

mvdbeek reviewed (approved):
Solid improvements, thanks a lot @wm75!

davelopez reviewed (commented):
Just a question, I noticed that we included `data_source_async` in the XSD documentation:

https://github.com/wm75/galaxy/blob/fd3a19afd2e8696cac805be1b5acbfd3e6e47708/lib/galaxy/tool_util/xsd/galaxy.xsd#L230

But I think is missing from the `ToolTypeType` definition in:

https://github.com/wm75/galaxy/blob/fd3a19afd2e8696cac805be1b5acbfd3e6e47708/lib/galaxy/tool_util/xsd/galaxy.xsd#L7152

We should probably add it, right?

wm75 committed:
Improve code comments and XSD

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -58,12 +58,17 @@ List of behavior changes associated with profile versions:
 ### 21.09
 
 - Do not strip leading and trailing whitespaces in `from_work_dir` attribute.
+- Do not use Galaxy python environment for `data_source` tools.
```
```suggestion
- Do not use Galaxy Python virtual environment for `data_source` tools. `data_source` tools should explicitly use the `galaxy-util` package.
```

davelopez commented on lib/galaxy/webapps/galaxy/controllers/async.py:
```diff
@@ -52,30 +50,62 @@ def index(self, trans, tool_id=None, data_secret=None, **kwd):
+            #
+            # we have an incoming data_id
+            #
             data = trans.sa_session.query(trans.model.HistoryDatasetAssociation).get(data_id)
```
Another thing we noticed is that this `data_id` is circulating unencoded between requests. It should be encoded right?

mvdbeek replied:
Ideally, but I guess this is somewhat internal ?

davelopez replied:
Yeah, it looks pretty internal, maybe when this gets translated to a proper API endpoint (if it should at some point) we can revisit the encoding.

wm75 committed:
Update lib/galaxy/tool_util/xsd/galaxy.xsd

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>
